### PR TITLE
test/k8sT: move creation / deletion of resources outside of `It` for tunnels test

### PR DESCRIPTION
**Pull request by ianvernon**
Signed-off by: Ian Vernon <ian@cilium.io>

Related-to: #3821

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by eloycoto** [comment]
I do not like this one. You can do in the same context without having two context here.  In both `afterEach` you can delete the manifest, an all in the same context no?

File: test/k8sT/Tunnels.go [hunk 1]
```diff
@@ -67,73 +67,103 @@ var _ = Describe("K8sValidatedTunnelTest", func() {
...
+			Expect(err).Should(BeNil())
+
+			_, err = kubectl.CiliumNodesWait()
+			Expect(err).Should(BeNil())
+
+			//Check that cilium detects a
+			By("Checking that BPF tunnels are in place")
+			status := kubectl.CiliumExec(ciliumPod, "cilium bpf tunnel list | wc -l")
+			status.ExpectSuccess()
+			if tunnels, _ := status.IntOutput(); tunnels != 4 {
```
Comment (eloycoto) [inline test/k8sT/Tunnels.go:151]: I think that this can be deleted, no?

**Review by eloycoto — CHANGES_REQUESTED** [review]

**Comment by ianvernon** [comment]
@eloycoto I don't see how this can be done in the same context. There are two different, disjoint sets of resources which are created for the two tests (vxlan Cilium DaemonSet, and the geneve Cilium Daemonset). This is precisely why I put them into two separate Contexts. If we want to add more tests to test Geneve vs. vxlan, we can put them in their according contexts. 

Ginkgo docs support what I say :
```
You should make use of Describe and Context containers to expressively organize the behavior of your code.
```

Source: https://onsi.github.io/ginkgo/#organizing-specs-with-containers-describe-and-context

I think this is a prime example of using `Context`.

File: test/k8sT/Tunnels.go (hunk 1 above)
Comment (ianvernon) [reply test/k8sT/Tunnels.go:151]: Can you clarify why? I don't understand.

Comment (ianvernon) [reply test/k8sT/Tunnels.go:151]: @eloycoto please provide clarification. Thanks!

**Comment by ianvernon** [comment]
test-me-please

File: test/k8sT/Tunnels.go (hunk 1 above)
Comment (eloycoto) [reply test/k8sT/Tunnels.go:151]: I'm not sure why I added the report if tunnels are != 4, but looks like not relevant for the test itself right now.

**Comment by eloycoto** [comment]
@ianvernon yes, it's a different context, but I would not add a Context for a single test. Things are complicated to debug *to not ginkgo people* when you have two context, here with two Context a new complexity layer is added, when test can be simpler doing it like this: 

```
var _ = Describe("K8sValidatedTunnelTest", func() {

	var kubectl *helpers.Kubectl
	var demoDSPath string
	var once sync.Once
	var logger *logrus.Entry

	BeforeAll(func() {
		logger = log.WithFields(logrus.Fields{"testName": "K8sTunnelTest"})
		logger.Info("Starting")

		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
		demoDSPath = kubectl.ManifestGet("demo_ds.yaml")
		kubectl.Exec("kubectl -n kube-system delete ds cilium")
		// Expect(res.Correct()).Should(BeTrue())

		waitToDeleteCilium(kubectl, logger)

		vxlanDSPath = kubectl.ManifestGet("cilium_ds.yaml")
		geneveDSPath = kubectl.ManifestGet("cilium_ds_geneve.yaml")
	})

	BeforeEach(func() {
		kubectl.NodeCleanMetadata()
		kubectl.Apply(demoDSPath)
	}, 600)

	AfterFailed(func() {
		kubectl.CiliumReport(helpers.KubeSystemNamespace, []string{
			"cilium bpf tunnel list",
			"cilium endpoint list"})
	})

	JustAfterEach(func() {
		kubectl.ValidateNoErrorsOnLogs(CurrentGinkgoTestDescription().Duration)
	})

	AfterEach(func() {
		kubectl.Delete(demoDSPath)

		_ = kubectl.Delete(vxlanDSPath)
		_ = kubectl.Delete(geneveDSPath)
		waitToDeleteCilium(kubectl, logger)

		err := kubectl.WaitCleanAllTerminatingPods()
		Expect(err).To(BeNil(), "Terminating containers are not deleted after timeout")
	})


	It("Check VXLAN mode", func() {
		res := kubectl.Apply(vxlanDSPath)
		res.ExpectSuccess("Unable to apply %s: %s", vxlanDSPath, res.CombineOutput())

		_, err := kubectl.WaitforPods(helpers.KubeSystemNamespace, "-l k8s-app=cilium", 500)
		Expect(err).Should(BeNil())

		ciliumPod, err := kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s1)
		Expect(err).Should(BeNil())

		_, err = kubectl.CiliumNodesWait()
		Expect(err).Should(BeNil())

		By("Checking that BPF tunnels are in place")
		status := kubectl.CiliumExec(ciliumPod, "cilium bpf tunnel list | wc -l")
		if tunnels, _ := status.IntOutput(); tunnels != 4 {
			cmds := []string{
				"cilium bpf tunnel list",
			}
			kubectl.CiliumReport(helpers.KubeSystemNamespace, cmds)
		}

		status.ExpectSuccess()
		Expect(status.IntOutput()).Should(Equal(5))

		By("Checking that BPF tunnels are working correctly")
		tunnStatus := isNodeNetworkingWorking(kubectl, "zgroup=testDS")
		Expect(tunnStatus).Should(BeTrue())
	}, 600)


	It("Check Geneve mode", func() {
		res := kubectl.Apply(geneveDSPath)
		res.ExpectSuccess("unable to apply %s: %s", geneveDSPath, res.CombineOutput())
		_, err := kubectl.WaitforPods(helpers.KubeSystemNamespace, "-l k8s-app=cilium", 500)
		Expect(err).Should(BeNil())

		ciliumPod, err := kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s1)
		Expect(err).Should(BeNil())

		_, err = kubectl.CiliumNodesWait()
		Expect(err).Should(BeNil())

		//Check that cilium detects a
		By("Checking that BPF tunnels are in place")
		status := kubectl.CiliumExec(ciliumPod, "cilium bpf tunnel list | wc -l")
		status.ExpectSuccess()
		if tunnels, _ := status.IntOutput(); tunnels != 4 {
			cmds := []string{
				"cilium bpf tunnel list",
			}
			kubectl.CiliumReport(helpers.KubeSystemNamespace, cmds)
		}
		Expect(status.IntOutput()).Should(Equal(5))

		By("Checking that BPF tunnels are working correctly")
		tunnStatus := isNodeNetworkingWorking(kubectl, "zgroup=testDS")
		Expect(tunnStatus).Should(BeTrue())
		//FIXME: Maybe added here a cilium bpf tunnel status?
	}, 600)
})
```
Feel free to merge, but I would like to keep test as simpler/small as possible. Finally with this behaviour the delete is outside of the IT, that it's what we need/want ;-)

**Comment by ianvernon** [comment]
@eloycoto responses inline.

> yes, it's a different context, but I would not add a Context for a single test.

Why not? I don't see a problem with adding a `Context` for a single test. `Context` provides a way for organizing code within a `Describe`. Besides, if we want to add more tests within these Contexts in the future, this change makes it easy to to do so. Furthermore, the change you propose moves the creation of resources to be within the `It`, which goes against the conventions I believe we have established:

* Creation of resources in a `BeforeAll` or a `BeforeEach`
* Testing of resources and testing state changes of those resources within `It`
* Deletion of resources in `AfterAll` or `AfterEach`

> Things are complicated to debug to not ginkgo people when you have two context, here with two Context a new complexity layer is added.

How is it complicated to debug?  What is this complexity layer? Please provide a justification.

**Comment by ianvernon** [comment]
@eloycoto can you reply to my above questions?

**Comment by eloycoto** [comment]
Hi @ianvernon 

Sorry, this was lost in the list of task. For me having Context for only one test make things complicated. For me is easy to debug when only one AfterEach, BeforeEach rather than debug 2 *Each functions, but it's a personal opinion. 

> @eloycoto responses inline.
>
>  yes, it's a different context, but I would not add a Context for a single test.

> Why not? I don't see a problem with adding a Context for a single test. Context provides a way for organizing code within a Describe. Besides, if we want to add more tests within these Contexts in the future, this change makes it easy to to do so. Furthermore, the change you propose moves the creation of resources to be within the It, which goes against the conventions I believe we have established:    

> Creation of resources in a BeforeAll or a BeforeEach
>     Testing of resources and testing state changes of those resources within It
>     Deletion of resources in AfterAll or AfterEach

That is a valid point, but in this case the resources are created in the test, and that is *ok*, we're testing that the Geneve/VXLan installation is working. 

As I said, your points are valid, I prefer to keep test without context when Context has only one test, but your points are valid, so I'm happy to merge this one ;-)

**Review by eloycoto — APPROVED** [review]

**Pull request merged**